parsl/executors/workqueue/exec_parsl_function.py

```diff
@@ -0,0 +1,183 @@
+
+
+def encode_function(user_namespace, fn, fn_name, fn_args, fn_kwargs):
+    # Returns a tuple (code, result_name)
+    # code can be exec in the user_namespace to produce result_name.
+    prefix = "parsl_"
+    args_name = prefix + "args"
+    kwargs_name = prefix + "kwargs"
+    result_name = prefix + "result"
+
+    # Add variables to the namespace to make function call
+    user_namespace.update({args_name: fn_args,
+                           kwargs_name: fn_kwargs,
+                           result_name: result_name})
+
+    if isinstance(fn, str):
+        code = encode_source_code_function(user_namespace, fn, fn_name, args_name, kwargs_name, result_name)
+    elif callable(fn):
+        code = encode_byte_code_function(user_namespace, fn, fn_name, args_name, kwargs_name, result_name)
+    else:
+        raise ValueError("Function object does not look like a function.")
+
+    return (code, result_name)
+
+
+def encode_source_code_function(user_namespace, fn, fn_name, args_name, kwargs_name, result_name):
+    # We drop the first line as it names the parsl decorator used (i.e., @python_app)
+    source = fn.split('\n')[1:]
+    fn_app = "{0} = {1}(*{2}, **{3})".format(result_name, fn_name, args_name, kwargs_name)
+
+    source.append(fn_app)
+
+    code = "\n".join(source)
+    return code
+
+
+def encode_byte_code_function(user_namespace, fn, fn_name, args_name, kwargs_name, result_name):
+    user_namespace.update({fn_name: fn})
+    code = "{0} = {1}(*{2}, **{3})".format(result_name, fn_name, args_name, kwargs_name)
+    return code
+
+
+def execute_function(map_file, function_file):
+    # Get all variables from the user namespace, and add __builtins__
+    user_ns = locals()
+    user_ns.update({'__builtins__': __builtins__})
+
+    function_info = load_pickled_file(function_file)
+
+    (fn, fn_name, fn_args, fn_kwargs) = unpack_function(function_info, user_ns)
+
+    mapping = load_pickled_file(map_file)
+    remap_all_files(mapping, fn_args, fn_kwargs)
+
+    (code, result_name) = encode_function(user_ns, fn, fn_name, fn_args, fn_kwargs)
+
+    exec(code, user_ns, user_ns)
+    result = user_ns.get(result_name)
+
+    return result
+
+
+if __name__ == "__main__":
+    try:
+        # parse the three required command line arguments:
+        # map_file: contains a pickled dictionary to map original names to
+        #           names at the execution site.
+        # function_file: contains the pickled parsl function to execute.
+        # result_file: any output (including exceptions) will be written to
+        #              this file.
+        try:
+            (map_file, function_file, result_file) = sys.argv[1:]
+        except ValueError:
+            print("Usage:\n\t{} function result mapping\n".format(sys.argv[0]))
+            raise
+
+        result = execute_function(map_file, function_file)
+    except Exception:
+        print("There was an error while setting up the function.")
```

Comment:
This exception clause also fires if there is a trigger in *executing* the function, not just in setting it up right?

Reply:
Fixed in 4098f35c9628f6e78f1f92b3180a50d6b060cb09. Exec and load error messages are now separated.